proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+    /// target on which the package plugin target depends. This function throws
+    /// an error if the tool cannot be found. The lookup is case sensitive.
+    public func tool(named name: String) throws -> Tool
+    
+    /// Information about a particular tool that is available to a plugin.
+    public struct Tool {
+        /// Name of the tool (suitable for display purposes).
+        public let name: String
+
+        /// Full path of the built or provided tool in the file system.
+        public let path: Path
+    }
+}
+```
+
+The `package` property is a reference to the top-level package to which the plugin is being applied. Through it, the entire subgraph of resolved packages reachable from it can be accessed.
+
+The properties related to looking up tools with a particular name are unchanged from the original SE-0303 proposal.
+
+This has the effect of factoring out all information related to the package and target, and it puts them into its own directed acyclic graph consisting of the following types:
+
+```swift
+/// Represents a single package in the graph (either the root or a dependency).
+public class Package {
+    /// The name of the package (for display purposes only).
+    public let name: String
+
+    /// The absolute path of the package directory in the local file system.
+    public let directory: Path
+
+    /// Any dependencies on other packages, in the same order as they are
+    /// specified in the package manifest.
+    public let dependencies: [Dependency]
+
+    /// Represents a resolved dependency of a package on another package. This is a
+    /// separate entity in order to make it easier for future versions of the API to
+    /// add information about the dependency itself.
+    public struct Dependency {
+        /// The package to which the dependency was resolved.
+        public let package: Package
+    }
+
+    /// Any regular products defined in this package (except plugin products),
+    /// in the same order as they are specified in the package manifest.
+    public let products: [Product]
+
+    /// Any regular targets defined in this package (except plugin targets),
+    /// in the same order as they are specified in the package manifest.
+    public let targets: [Target]
+}
+
+/// Represents a single product defined in a package. Specializations represent
+/// different types of products.
+public class Product {
+    /// The name of the product, as defined in the package manifest. It's unique
+    /// among the products of the package in which it is defined.
+    public let name: String
+    
+    /// The targets that directly comprise the product, in the order in which
+    /// they are declared in the package manifest. The product will contain the
+    /// transitive closure of the these targets and their depdendencies.
+    public let targets: [Target]
+}
+
+/// Represents an executable product defined in a package.
+public class ExecutableProduct: Product {
```

Comment:
could we avoid class inheritance here? maybe `Product` can be a protocol and the concrete product types structs?

Reply:
That makes sense, though I’d note that these objects do have identity semantics and not value semantics, since they’re part of a DAG.  But it might still make sense to implement them as a protocol and structs, even with identity semantics.

The main thing I wanted was to be able to have type-specific properties and not have to have some separate `info` struct for each kind, and also to keep the types as open-ended as possible here since we know there will be a wider set of them in the future.

For users of this API, does it matter if it’s implemented using inheritance?  Since plugins won’t be defining new subclasses here of their own?

Reply:
In the draft implementation I've changed `Product` and `Target` to be protocols, with the specific types of each being structs.  For graph traversal we do still need some way of uniquely identifying each entity, e.g. using their names or some other kind of value-based identifier.  I did have some difficulty making these protocols conform to `Identifiable` since it resulted in 'self-or-associated-type" requirements when trying to then use the protocols as existentials (e.g. in `products: [Product]`.  But using the an identifier formed from the package identity and the name might work.

I'm not really sure why the DAG nodes need to be value types though.  It would be easier to just use `ObjectIdentifier` for these objects.  But we can at least get rid of inheritance.